File: src/main/java/com/polygloat/security/controllers/InvitationController.java
package com.polygloat.security.controllers;

import com.polygloat.dtos.response.InvitationDTO;
import com.polygloat.exceptions.NotFoundException;
import com.polygloat.model.Invitation;
import com.polygloat.model.Permission;
import com.polygloat.model.Repository;
import com.polygloat.service.InvitationService;
import com.polygloat.service.RepositoryService;
import com.polygloat.service.SecurityService;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.LinkedHashSet;
import java.util.Set;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/invitation")
@RequiredArgsConstructor(onConstructor = @__(@Autowired))
public class InvitationController {

    private final InvitationService invitationService;
    private final SecurityService securityService;
    private final RepositoryService repositoryService;

    @GetMapping("/accept/{code}")
    public void acceptInvitation(@PathVariable("code") String code) {
        invitationService.removeExpired();
        invitationService.accept(code);
    }

    @GetMapping("/list/{repositoryId}")
    public Set<InvitationDTO> getRepositoryInvitations(@PathVariable("repositoryId") Long id) {
        Repository repository = repositoryService.findById(id).orElseThrow(NotFoundException::new);
        securityService.checkRepositoryPermission(id, Permission.RepositoryPermissionType.MANAGE);
        return invitationService.getForRepository(repository).stream().map(InvitationDTO::fromEntity).collect(Collectors.toCollection(LinkedHashSet::new));
    }

    @DeleteMapping("/{invitationId}")
    public void deleteInvitation(@PathVariable("invitationId") Long id) {
        Invitation invitation = invitationService.findById(id).orElseThrow(NotFoundException::new);
        securityService.checkRepositoryPermission(invitation.getPermission().getRepository().getId(), Permission.RepositoryPermissionType.MANAGE);
        invitationService.delete(invitation);
    }
}

File: src/main/java/io/tolgee/security/controllers/InvitationController.kt
package io.tolgee.security.controllers

import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import io.tolgee.dtos.response.InvitationDTO
import io.tolgee.exceptions.NotFoundException
import io.tolgee.model.Invitation
import io.tolgee.model.Permission
import io.tolgee.service.InvitationService
import io.tolgee.service.OrganizationRoleService
import io.tolgee.service.ProjectService
import io.tolgee.service.SecurityService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.stream.Collectors

@RestController
@RequestMapping("/api/invitation")
@Tag(name = "User invitations to project")
open class InvitationController @Autowired constructor(
        private val invitationService: InvitationService,
        private val securityService: SecurityService,
        private val projectService: ProjectService,
        private val organizationRoleService: OrganizationRoleService
) {
    @GetMapping("/accept/{code}")
    @Operation(summary = "Accepts invitation to project")
    open fun acceptInvitation(@PathVariable("code") code: String?): ResponseEntity<Void> {
        invitationService.removeExpired()
        invitationService.accept(code)
        return ResponseEntity(HttpStatus.OK)
    }

    @GetMapping("/list/{projectId}")
    @Operation(summary = "Prints all invitations to project")
    open fun getRepositoryInvitations(@PathVariable("projectId") id: Long): Set<InvitationDTO> {
        val repository = projectService.get(id).orElseThrow { NotFoundException() }!!
        securityService.checkRepositoryPermission(id, Permission.ProjectPermissionType.MANAGE)
        return invitationService.getForRepository(repository).stream().map { invitation: Invitation? ->
            InvitationDTO.fromEntity(invitation)
        }.collect(Collectors.toCollection { LinkedHashSet() })
    }

    @DeleteMapping("/{invitationId}")
    @Operation(summary = "Deletes invitation by ID")
    open fun deleteInvitation(@PathVariable("invitationId") id: Long): ResponseEntity<Void> {
        val invitation = invitationService.findById(id).orElseThrow {
            NotFoundException()
        }
        invitation.permission?.let {
            securityService.checkRepositoryPermission(invitation.permission!!.project!!.id,
                    Permission.ProjectPermissionType.MANAGE)
        }

        invitation.organizationRole?.let {
            organizationRoleService.checkUserIsOwner(invitation.organizationRole!!.organization!!.id!!)
        }

        invitationService.delete(invitation)
        return ResponseEntity(HttpStatus.OK)
    }
}
